yarn/history/src/main/scala/org/apache/spark/deploy/history/yarn/YarnHistoryService.scala

```diff
@@ -0,0 +1,981 @@
+  }
+
+  /**
+   * Stop the service; this triggers flushing the queue and, if not already processed,
+   * a pushing out of an application end event.
+   *
+   * This operation will block for up to `maxTimeToWaitOnShutdown` milliseconds
+   * to await the asynchronous action queue completing.
+   */
+  override protected def serviceStop {
+    // if the queue is live
+    if (!stopped.get) {
+
+      if (!appEndEventProcessed.get()) {
+        // push out an application stop event if none has been received
+        logDebug("Generating a SparkListenerApplicationEnd during service stop()")
+        val current = now()
+        innerEnqueue(new HandleSparkEvent(SparkListenerApplicationEnd(current), current))
+      }
+
+      // flush
+      logInfo(s"Shutting down: pushing out ${actionQueue.size} events")
+      innerEnqueue(FlushTimelineEvents())
+
+      // stop operation
+      postThreadActive synchronized {
+        innerEnqueue(StopQueue())
+        // now await that marker flag
+        postThreadActive.wait(maxTimeToWaitOnShutdown)
+      }
+
+      if (!actionQueue.isEmpty) {
+        // likely cause is ATS not responding.
+        // interrupt the thread, albeit at the risk of app stop events
+        // being lost. There's not much else that can be done if
+        // ATS is being unresponsive
+        logWarning(s"Did not finish flushing actionQueue before " +
+          s"stopping ATSService, eventQueueBacklog= ${actionQueue.size}")
+        eventHandlingThread.foreach(_.interrupt())
+      }
+      stopTimelineClient
+      logInfo(s"Stopped: $this")
+    }
+  }
+
+  /**
+   * Get the current entity, creating one on demand if needed, setting up its entity Id
+   * from the current appID and attempt ID values, filters from the current name and user
+   * @return the current entity
+   */
+  private def getOrCreateCurrentEntity() = {
+    curEntity.getOrElse {
+      currentEventCount = 0
+      val entity = createTimelineEntity(
+        _applicationId,
+        _attemptId,
+        sparkApplicationId,
+        sparkApplicationAttemptId,
+        appStartEventProcessed.get(),
+        appName,
+        userName)
+      curEntity = Some(entity)
+      logDebug(s"Demand creation of new entity ${describeEntity(entity)}")
+      entity
+    }
+  }
+
+  /**
+   * Perform any preflight checks.
+   * @param entity timeline entity to review.
+   */
+  private[yarn] def preflightCheck(entity: TimelineEntity): Unit = {
+    require(entity.getStartTime != null,
+      s"No start time in ${describeEntity(entity)}")
+  }
+
+  /**
+   * If there is any available entity to be sent, push to timeline server
+   */
+  private def flushEntity(): Unit = {
+    if (entityList.nonEmpty) {
+      val count = flushCount.incrementAndGet()
+      logDebug(s"flushEntity #$count: list size ${entityList.size}")
+      var client = getTimelineClient
+      val lastUpdated = now()
+      // attempt to post each entity, removing from the list all of those
```

Comment:
In my original, unfinished code ([1]), I used the local fs to buffer these events, because I wasn't really comfortable with the ATS becoming a bottleneck and causing memory pressure on the driver. That allowed the code to upload events always in order while avoiding caching everything in memory.

It might be a little paranoid, but it does avoid both issues. And if the ATS is really borked, well, the app's event log would eventually be lost (or left incomplete).


[1] https://github.com/vanzin/spark/blob/yarn-timeline/yarn/timeline/src/main/scala/org/apache/spark/deploy/yarn/timeline/YarnTimelineClientImpl.scala